### ISSUE TITLE
bibtexparser does not handle files from bibutils xml2bib

### ISSUE THREAD

**Issue by humm** [pubs/pubs#23]
Might not be our fault.

xml2bib generate files that have a different first character. `cmp` give that output: 

'''
test2.bib test3.bib differ: byte 1, line 1 is 357 � 100 @
'''

To investigate further.

**Comment by omangin** [comment]
Actually if this is a bibtexparser bug, it should be reported upstream instead because bibtexparser is supposed to hande BOM. Actually it does so for `0xEF 0xBB 0xBF` (<https://github.com/sciunto-org/python-bibtexparser/blob/master/bibtexparser/bparser.py#L103-L105>). What is `0xFE 0xFF` supposed to be for? Isn't it for UTF-16?

However, according to <https://docs.python.org/3/howto/unicode.html>:
> The Unicode character U+FEFF is used as a byte-order mark (BOM), and is often written as the first character of a file in order to assist with autodetection of the file’s byte ordering. Some encodings, such as UTF-16, expect a BOM to be present at the start of a file; when such an encoding is used, the BOM will be automatically written as the first character and will be silently dropped when the file is read. There are variants of these encodings, such as ‘utf-16-le’ and ‘utf-16-be’ for little-endian and big-endian encodings, that specify one particular byte ordering and don’t skip the BOM.

So actually this might be handled correctly by the `io.open` function. In which case the error is caused by <https://github.com/pubs/pubs/blob/b2ebc67f5e028a5dc1633d94397e7c8bd594bb08/pubs/content.py#L40-L44> that forces the openning in UTF-8.

Can you `file` your file? What is your default locale? Does the error persist without the fix, if instead you remove the encoding='utf-8' in <https://github.com/pubs/pubs/blame/b2ebc67f5e028a5dc1633d94397e7c8bd594bb08/pubs/content.py#L42>?

**Comment by humm** [comment]
Reopening a closed issues about dark aspects of encodings... you are looking for that much problem this early in the year ? :)

I went around and did some testing when investigating that bug. The thing is, the xml2bib does include the BOM by default, but has an option to not include it in the file. The produced file is UTF-8. See for yourself, the file is in bibexamples:

    >> file -I bibutils.bib
    bibutils.bib: text/plain; charset=utf-8

Frankly, it might be a bibtexparser bug, feel free to upstream the thing.

The thing is, besides all that, you don't address the real question: what is wrong with my fix? And should we really focus on this issue? This previously *closed* issue?

**Comment by humm** [comment]
This issue warrants a new label all on its own. :)

**Comment by omangin** [comment]
Nothing is wrong with your fix except that in my opinion this should not be fixed. From what you say it seems that this is a bug of xml2bib adding a BOM it should not be adding. So there is no reason to add code to pubs that is meant at fixing other programs bug.

**Comment by humm** [comment]
First, here is bib2xml help:
```
$ xml2bib --help
xml2bib, bibutils suite version 5.6 date 2015-01-26
Converts the MODS XML intermediate reference file into Bibtex

usage: xml2bib xml_file > bibtex_file

  xml_file can be replaced with file list or omitted to use as a filter

  -h,  --help               display this help
  -v,  --version            display version
  -at, --abbreviatedtitles  use abbreviated titles, if available
  -fc, --finalcomma         add final comman to bibtex output
  -sd, --singledash         use one dash '-', not two '--', in page ranges
  -b,  --brackets           use brackets, not quotation marks surrounding data
  -w,  --whitespace         use beautifying whitespace to output
  -sk, --strictkey          use only alphanumeric characters for bibtex key
                            (overly strict, but useful for other programs)
  -nl, --no-latex           no latex encodings; put characters in directly
  -nb, --no-bom             do not write Byte Order Mark in UTF8 output
  -U,  --uppercase          write bibtex tags/types in upper case
  -s,  --single-refperfile  one reference per output file
  -i, --input-encoding      interpret input file with requested character set
                            (use argument for current list)
  -o, --output-encoding     write output file with requested character set
                            (use argument for current list)
  --verbose                 for verbose
  --debug                   for debug output

Citation codes generated from <REFNUM> tag.   See
http://sourceforge.net/p/bibutils/home/Bibutils for more details
```
So `xml2bib` will write the BOM, while `xml2bib -un` will not. It is not a bug, it is a choice for the default behaviour.

Second, I figured it out. This is bibtexparser's bug.

**Comment by humm** [comment]
First, let's clarify: `0xEF 0xBB 0xBF` is the utf-8 representation of the *unicode* character `U+FEFF` (*not* `0xFE 0xFF`). 

All right, let's open a Python 3 shell:
```
>>> '\xef\xbb\xbf' == '\ufeff'
False
```
Ah ah! Let's look into it:
```
>>> len('\xef\xbb\xbf')
3
>>> len('\ufeff')
1
```
In fact, the correct code is:
```
>>> bytes('\ufeff', 'utf-8') == b'\xef\xbb\xbf'
True
```
Because `bytes` and `str` are different things in Python 3. And bibtexparser, not only uses `str` in Python 3 like the `unicode` function in Python 2, but also checks the *three* first characters, which is find for bytes string, but in unicode, the BOM is just one character. 

In conclusion: my code is sooooo superior, and yes, I will do a pull request to fix their mess.

**Comment by humm** [comment]
BTW, I do think that leaving those 2 lines there are a good defence against bibtexparser potential future bugs with the BOM. We're basically sending the same data to bibtexparser, but in a simpler way (the BOM is *useless* in utf-8). So, I am closing this. If you disagree, reopen it.
